### DISCUSSION TITLE
Lighting toggle in editor (at least for 2D)

### DISCUSSION THREAD

**Discussion by Meorge** [godotengine/godot-proposals#9578] [Editor]
I've been enjoying working on my 2D game in Godot, but I've run into a bit of an annoyance with the lighting system. I'd like my areas to be dark by default during gameplay, and have them be lit up by objects in the world, such as the player's flashlight. However, when editing the areas, I'd like to be able to clearly see everything.

Currently, when I have my in-game negative light enabled, my scene looks like this:
![CleanShot 2024-04-22 at 18 57 45@2x](https://github.com/godotengine/godot-proposals/assets/9957987/740f9f11-3cd7-4805-9af5-0014b170d700)

In order to see the level effectively, I disable my node containing the WorldEnvironment and negative light:
![CleanShot 2024-04-22 at 18 57 58@2x](https://github.com/godotengine/godot-proposals/assets/9957987/5a090974-bfaa-4d78-95db-4b6519474fd4)

The trouble with this is that I have to remember to re-enable it before I run the game to see what it will actually look like with the correct lighting. I've also probably pushed a bunch of commits that inadvertently disabled and re-enabled the node as I forgot and remembered that it was modified.

It'd be great if there was an option along the top row above the scene view, where we could toggle lights on and off.

![CleanShot 2024-04-22 at 19 04 54@2x](https://github.com/godotengine/godot-proposals/assets/9957987/31ed7224-cc27-4eeb-8eb4-a210b35c0ee4)

(If there is already some way to do this, please let me know - it'd be great to hear! But I feel like I've searched pretty well for it and haven't been able to find anything.)

**Comment by aXu-AP** [comment]
Workaround could be to disable it in editor and add a script which enables it runtime.
I think I saw a similiar proposal for 3D some time ago, but I wasn't able to find it now...

**Reply by Meorge** [reply]
Someone also proposed this in a Discord chat. While I think it _would_ more or less _work_ as an interim solution, it would also make it more difficult to preview how the whole level would look with lighting on. Generally, it just feels like this kind of thing shouldn't need a workaround.

I am one of the many who came to Godot from Unity, and I know there they had a button in the toolbar above the scene view that allowed you to quickly and easily toggle lighting within the scene view, without impacting the lights in the game itself.

**Comment by aXu-AP** [comment]
I cobbled together a plugin to solve this problem: [aXu-AP/LightToggle2D](https://github.com/aXu-AP/LightToggle2D)
Do note that it isn't fully tested so do a backup before trying it out 😅 
![kuva](https://github.com/godotengine/godot-proposals/assets/1621768/0b64163c-80fd-43d4-95dc-2997c7f20040)

**Reply by Meorge** [reply]
Woah, thanks! I'll definitely have to see how this works with my project. Perhaps if it seems to be working well, it (and/or its underlying principles) could be used as a foundation for a more permanent implementation in the engine?

**Reply by aXu-AP** [reply]
I wouldn't use it as a base for core since it is quite hacky: I didn't find a way to disable light rendering, nor did I find a way to temporarily disable light objects (their state would get saved with the scene). The trick I use in the plugin is to create hidden copies of existing lights but with negating effect (ie. switch add and subtract blend modes). It works, but on light-heavy scenes it might have performance impact in editor. This approach also doesn't work with CanvasModulate, which is often used for darkening the ambience (but since you seem to use DirectionalLight2D for that, it shouldn't be an issue).

For an actual implementation in engine, I would either disable lighting pass or add a bit of code in Light2D and CanvasModulate to disable it when needed. Maybe look into how debug views are implemented in 3D view, solution could be something similiar.
